Update istio-remote helm chart to support automatic sidecar injection

Rigdon opened the pull request:
WIP, this needs further testing.
This removes the local Istio service endpoints in favor of addressing remote endpoints directly and provides a basic setup for the Citadel and SIdecar Injectors. I will update the README with instructions if we agree on this approach and it works for others.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6366#" title="Author self-approved">Rigdon</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **andraxylia**

Assign the PR to them by writing `/assign @andraxylia` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[install/OWNERS](https://github.com/istio/istio/blob/master/install/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["andraxylia"]} -->

istio-testing commented:
Hi @Rigdon. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

Rigdon commented:
@tiswanso I migrated our internal Helm chart into the istio-remote directory. This is pretty much exactly what I deployed and the sidecar injector just worked. As I mentioned in the working group meeting I was unable to get a full end to end test due to #6313 but that's unrelated to multi-cluster. We're going to create a couple clusters without custom domains and I'll keep working at testing this.

I removed the endpoint and service templates in favor of using the remote endpoints directly. I could see setting up `ExternalName` type services though if that were preferable.

Rigdon commented:
Fixes #6092 
Fixes #6321

gyliu513 commented on install/kubernetes/helm/istio-remote/charts/security/templates/deployment.yaml:
```diff
@@ -29,13 +31,18 @@ spec:
             - --grpc-hostname=citadel
             - --citadel-storage-namespace={{ .Release.Namespace }}
+          {{- if .Values.selfSigned }}
+            - --self-signed-ca=true
```
Does the two cluster can communicate with each other if using different root cert?

sdake replied:
@gyliu513 this looks like a copy and paste from the main helm chart.  I think that is fine for now.

Rigdon replied:
yeah, what Steve said. The security and sidecar injector charts are exact copies of the subcharts in `install/kubernetes/helm/istio`

gyliu513 replied:
But my question is does mTLS works cross clusters for such case?

tiswanso replied:
@gyliu513  -- AFAIK, yes, as long as each cluster's root cert was signed by the same CA.

gyliu513 commented on install/kubernetes/helm/istio-remote/charts/sidecarInjectorWebhook/Chart.yaml:
```diff
@@ -0,0 +1,11 @@
+tillerVersion: ">=2.7.2"
+description: Helm chart for sidecar injector webhook deployment
+keywords:
+  - istio
```
Please add `security` as well.

sdake replied:
@gyliu513 I prefer not to carry deltas in the istio-remote chart.  The update model should be a simple copy and paste.  that said, if you think security belongs as a keyword, might submit that to the main istio helm chart.

Rigdon replied:
It is a straight copy/paste. I'd personally prefer to use `helm dependency` but that's not a pattern that is currently in use for the project. I could change it to symlinks if we think t's a better approach to avoid drift but I generally don't like them in git repos.

tiswanso replied:
@sdake @gyliu513 --what do you guys think?  IMO, relative path symlinks are ok and easier to review--explicitly ensuring "there's no change here".

gyliu513 replied:
+1 to symlinks.

gyliu513 commented on install/kubernetes/helm/istio-remote/charts/sidecarInjectorWebhook/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,21 @@
+  name: istio-sidecar-injector-{{ .Release.Namespace }}
+subjects:
+  - kind: ServiceAccount
+{{- if .Values.global.rbacEnabled }}
```
I was a bit confused, seems here the rbac logic in your PR only handling the serviceAccount name? 

Same issue for the local chart. @ymesika as you initiated the original PR for local chart, any comments for this?

sdake replied:
I don't think its helpful to fix bugs in the main chart by reviewing istio-remote :)  It may make more sense to just fix those problems independently and use a copy and paste as was done here (and was done in the past for citadel).

gyliu513 replied:
make sense.

gyliu513 commented on install/kubernetes/helm/istio-remote/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,8 @@
+    name: istio-reader
+  rules:
+    - apiGroups: ['']
+      resources: ['nodes', 'pods', 'services', 'endpoints']
```
Do we need to add permission for gateway, destinationrules, virtualservices?

Rigdon replied:
No because those resources will only exist in the primary control plane cluster.

tiswanso replied:
@gyliu513  - I confirm  @Rigdon's answer

gyliu513 replied:
yes, we have only one control panel.

gyliu513 commented on install/kubernetes/helm/istio-remote/templates/configmap.yaml:
```diff
@@ -0,0 +1,114 @@
+    {{- if .Values.global.proxy.envoyStatsd.enabled }}
+      #
+      # Statsd metrics collector converts statsd metrics into Prometheus metrics.
+      statsdUdpAddress: {{ .Values.global.proxy.envoyStatsd.host }}:{{ .Values.global.proxy.envoyStatsd.port }}
```
namespace also needed here

Rigdon replied:
This is from the main chart as well. If the namespace is required it should be in`{{ .Values.global.proxy.envoyStatsd.host }}`

gyliu513 replied:
main chart has a bug and has been fixed at https://github.com/istio/istio/pull/6349

gyliu513 commented on install/kubernetes/helm/istio-remote/templates/endpoints.yaml:
```diff
@@ -1,113 +0,0 @@
-apiVersion: v1
```
If removing this, then how to let the local istio connect to remote istio?

BTW: I also created a PR to add more checking for this at https://github.com/istio/istio/pull/6393

sdake replied:
agreed, I don't understand how this works properly.  @Rigdon could you expand on how one deploys Istio in this setup?

Rigdon replied:
The proxy containers in the remote cluster will use the remote endpoints directly as this is written. We could add some `ExternalName` services so that they are referencing local service names but I couldn't think of any immediate value in that.

Rigdon replied:
I can go ahead and update the README with details if we think this is generally the right direction.

sdake replied:
The README.md should not exist - it was removed once and replaced with a link to istio.io.  No sense updating README.md, although istio.io will need any updates that are relevant to changes made in the charts.

Rigdon replied:
👍

tiswanso replied:
@sdake @Rigdon @gyliu513  -- does removing this mandate using auto-sidecar injection?  (or pretty much mandate it?)

With the selectorless service + endpoint setup on the remote, we're able to make the manual injection workflow use the exact same `kube-inject` output for apps as if they were deployed on the cluster with the Istio control-plane.  IMO, this is fairly nice.  

However, I think a similar workflow could still work with this change if the `istioctl kube-inject` is run against a remote cluster with the `--meshConfigMapName` (or maybe `--injectConfigMapName`)  option pointing to the `istio` configmap in this change.  I hope to get time to play with that scenario.

Rigdon replied:
Yeah I guess I was hoping (I haven't been using manual injection at all) that `istioctl kube-inject` would honor the meshConfig in the target cluster. If that's not the case then adding the `ExternalName` type service may be a better option.

gyliu513 replied:
thanks all, I think this is good to remove this as long as we can make sure manual injection works. @sdake @tiswanso @Rigdon

sdake replied:
@Rigdon @tiswanso I think the outstanding question that needs answered prior to merging this PR is whether or not manual injection works with it.  All of the e2e tests for multicluster depend on manual injection, and while i'm not opposed to changing those to automatic injection, we don't want to break the test cases with this merge.

Further we do need manual injection to work - atleast for 1.0 and until automatic injection has less tread on the tires.

Cheers
-steve

Rigdon commented on install/kubernetes/helm/istio-remote/templates/namespace.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
👍

sdake commented on install/kubernetes/helm/istio-remote/values.yaml:
```diff
@@ -97,6 +90,14 @@ global:
 
+  # Remote Istio endpoints. Can be hostnames or IP addresses
+  # The Pilot address is required. The others are optional.
+  remotePilotAddress: ""
```
Kubernetes applications won't work in this model unless there is some magic I don't understand.  Currently we specify the pilotendpoint so that the sidecars inside kubernetes can find pilot and receive the mesh calculation.

Could you add more detail as to how the deployment works in this situation?

Rigdon replied:
I use `remotePilotAddress` as the discoveryAddress [here](https://github.com/istio/istio/pull/6366/files#diff-4e2f8fda81dd49dc86533e3c3af59d8bR106) in the mesh config (along with the other `remoteX` addresses. The proxy sidecars will then use those adresses instead of local default service names.

sdake replied:
Sounds like an improvement.  Mind I test prior to merge?  cc @baodongli @tiswanso

Rigdon replied:
Please!

sdake replied:
@Rigdon are you using proyv1, or v2?  v1 is deprecated, and IIUC will disappear in the 1.0 release.  We should be marching towards that model if possible.

Rigdon replied:
The default in both charts now is the `proxyv2` image, though as @tiswanso pointed out the sidecars are still configured by default to use the old http ports rather than the newer grpc ports. I tried to keep as much the same as possible between the two charts.

tiswanso replied:
@sdake -- I think we want to fix this in the main charts ASAP.  From your slack msg, I'm guessing you're looking at it.  Do you know if we have an issue already for tracking the latest xDS config stuff the helm charts?

@baodongli  ^^

Rigdon commented on install/kubernetes/helm/istio-remote/values.yaml:
```diff
@@ -1,44 +1,103 @@
-# Use --set to specify the endpoints of the main Istio control plane
-# NB: These defaults do not work properly - they must be overriden with
-# IP addresses
+# Use --set or additional values.yaml file to configure settings.
```
I'll go through and review the variables again. I removed all of the sections related to Istio components that won't be installed on the remote clusters.

Rigdon commented:
@sdake I removed the namespace and any variables in the values.yaml that aren't used by the security and injector subcharts

tiswanso commented on install/kubernetes/helm/istio-remote/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,14 @@
+  name: istio-reader
+subjects:
+- kind: ServiceAccount
+  name: istio-control-plane-service-account
```
This service account isn't created in any chart that I can see.  Did you add a new file to create this service account but it didn't get added to changeset?

Rigdon commented:
that was an error on my part. It's fixed now.

On Wed, Jun 20, 2018 at 3:40 PM Tim Swanson <notifications@github.com>
wrote:

> *@tiswanso* commented on this pull request.
> ------------------------------
>
> In install/kubernetes/helm/istio-remote/templates/clusterrolebinding.yaml
> <https://github.com/istio/istio/pull/6366#discussion_r196917651>:
>
> > @@ -0,0 +1,14 @@
> +apiVersion: rbac.authorization.k8s.io/v1
> +kind: ClusterRoleBinding
> +metadata:
> +  name: istio-multi
> +  labels:
> +    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
> +roleRef:
> +  apiGroup: rbac.authorization.k8s.io
> +  kind: ClusterRole
> +  name: istio-reader
> +subjects:
> +- kind: ServiceAccount
> +  name: istio-control-plane-service-account
>
> This service account isn't created in any chart that I can see. Did you
> add a new file to create this service account but it didn't get added to
> changeset?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/6366#pullrequestreview-130558699>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABcyuMBAzVSZLAMb09qvqgWypVvxLsF4ks5t-qU2gaJpZM4UsYlZ>
> .
>


-- 
Ryan Rigdon

tiswanso commented on install/kubernetes/helm/istio-remote/templates/serviceaccount.yaml:
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: istio-multi
```
you want `istio-control-plane-service-account` instead of `istio-multi`

Rigdon replied:
I changed it the other way. Kept the `istio-multi` SA name to match the docs you'd written and fixed the role binding instead.

tiswanso commented on install/kubernetes/helm/istio-remote/templates/configmap.yaml:
```diff
@@ -0,0 +1,114 @@
+      controlPlaneAuthPolicy: MUTUAL_TLS
+      #
+      # Address where istio Pilot service is running
+      discoveryAddress: {{ .Values.global.remotePilotAddress }}:15005
```
We need to fix this section for envoyv1 xDS v. envoyv2 xDS (grpc).  ports 15005 and 15007 are the envoy v1 (legacy) pilot endpoints.  The envoy v2 (grpc) endpoints are 15010 and 15011.

Rigdon replied:
yeah... I noticed that. Didn't want to deviate from the [main chart](https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/templates/configmap.yaml#L85-L99) though.

tiswanso replied:
Cool... good point.  Not sure if there's an issue already for this.

gyliu513 replied:
There is an issue here https://github.com/istio/istio/pull/6452

gyliu513 commented on install/kubernetes/helm/istio-remote/templates/configmap.yaml:
```diff
@@ -0,0 +1,114 @@
+    {{- if .Values.global.proxy.envoyStatsd.enabled }}
+      #
+      # Statsd metrics collector converts statsd metrics into Prometheus metrics.
+      statsdUdpAddress: {{ .Values.global.proxy.envoyStatsd.host }}:{{ .Values.global.proxy.envoyStatsd.port }}
```
Should not here be remote statsd?

Rigdon replied:
the existing statsd configuration in values.yaml is adequate. the statsd collector could be local or remote and either is supported by the current configuration options.

tiswanso commented:
@Rigdon  -- hey, would you mind if we move the service-account stuff out into another PR to maybe get it in faster?  I might end up needing it to cleanup the multicluster e2e tests.

Rigdon commented on install/kubernetes/helm/istio-remote/charts/sidecarInjectorWebhook:
```diff
@@ -0,0 +1 @@
+../../istio/charts/sidecarInjectorWebhook
```
symlink

Rigdon commented on install/kubernetes/helm/istio-remote/charts/security:
```diff
@@ -0,0 +1 @@
+../../istio/charts/security
```
symlink

Rigdon commented:
@sdake @tiswanso @gyliu513 

I updated this with all of the additional feedback. The big things being the switch to symlinks for the security and injector charts and I updated the readme to point at the preliminary docs page. 

I'll also pull the service account parts into a separate PR so that we can get them merged for Tim. I'll leave them here for now too in case this gets merged quickly and fix the conflict Monday if not.

> hey, would you mind if we move the service-account stuff out into another PR to maybe get it in faster? I might end up needing it to cleanup the multicluster e2e tests.

istio-testing commented:
@Rigdon: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

Rigdon commented:
The commands that appear to be failing in both the lint and e2e-galley tests are working fine in Helm version 2.8.2.

## e2e-galley
```
# test
$ helm template --namespace=istio-system install/kubernetes/helm/istio-remote
Error: error reading charts/security: read /go/src/istio.io/istio/install/kubernetes/helm/istio-remote/charts/security: is a directory

#local 
$ helm template --namespace=istio-system install/kubernetes/helm/istio-remote > /dev/null
$ echo $?
0
```

## Lint
```
# test
$ helm lint ./install/kubernetes/helm/istio ./install/kubernetes/helm/istio-remote
==> Linting ./install/kubernetes/helm/istio
Lint OK

==> Linting ./install/kubernetes/helm/istio-remote
[ERROR] templates/: error reading charts/security: read /go/src/istio.io/istio/install/kubernetes/helm/istio-remote/charts/security: is a directory


# local 
$ helm lint ./install/kubernetes/helm/istio ./install/kubernetes/helm/istio-remote
==> Linting ./install/kubernetes/helm/istio
Lint OK

==> Linting ./install/kubernetes/helm/istio-remote
Lint OK

2 chart(s) linted, no failures
```

It looks like an issue with the symlinks.

gyliu513 commented:
/ok-to-test

gyliu513 commented:
@Rigdon it is weird that the symlinks failed, if this is the case, I think we may need to open an issue in helm community and have some discussion there?

Rigdon commented:
As I mentioned above it works with the version of Helm I'm using. I expect
that it's that the istio project is using an older version of Helm or the
symlinks aren't being cloned correctly. I agree with @sdake that we should
move on without the symlinks for now. Hopefully we can just use helm
dependency in the near future.

On Mon, Jun 25, 2018, 7:54 PM Guang Ya Liu <notifications@github.com> wrote:

> @Rigdon <https://github.com/Rigdon> it is weird that the symlinks failed,
> if this is the case, I think we may need to open an issue in helm community
> and have some discussion there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/6366#issuecomment-400131415>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABcyuNRzA43jcTb2F0iBmqVXKIxj9CGOks5uAXg5gaJpZM4UsYlZ>
> .
>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6366#pullrequestreview-131847348" title="LGTM">gyliu513</a>*, *<a href="https://github.com/istio/istio/pull/6366#" title="Author self-approved">Rigdon</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[install/OWNERS](https://github.com/istio/istio/blob/master/install/OWNERS)~~ [gyliu513]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

gyliu513 commented:
@Rigdon can you rebase so that I can do some test for this PR?

istio-testing commented:
@Rigdon: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
New changes are detected. LGTM label has been removed.

istio-testing commented:
@Rigdon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 4feddc66b7ee32c7fb2fd10a64d6d09d650b585c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-unit-tests/9497/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Rigdon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 4feddc66b7ee32c7fb2fd10a64d6d09d650b585c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-unit-tests/9497/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | 4feddc66b7ee32c7fb2fd10a64d6d09d650b585c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-presubmit/13708/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

gyliu513 commented:
@Rigdon please note that we do not need `.Values.global.rbacEnabled` anymore,, as RBAC will always enabled, checkout here https://github.com/istio/istio/pull/6657 for detail.

/cc @sdake

gyliu513 commented on install/kubernetes/helm/istio-remote/charts/security/templates/cleanup-old-ca.yaml:
```diff
@@ -0,0 +1,41 @@
+        app: {{ template "security.name" . }}
+        release: {{ .Release.Name }}
+    spec:
+{{- if .Values.global.rbacEnabled }}
```
Those are not needed anymore #6657

tiswanso commented on install/kubernetes/helm/istio-remote/charts/security/templates/clusterrolebinding.yaml:
```diff
@@ -13,5 +13,34 @@ roleRef:
+{{- else }}
+    name: default
+{{- end }}
     namespace: {{ .Release.Namespace }}
```
missing `{{ end }}` for `{{- if $.Values.cleanUpOldCA }}`

istio-testing commented:
@Rigdon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 4feddc66b7ee32c7fb2fd10a64d6d09d650b585c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-presubmit/13708/) | `/test istio-presubmit`
prow/istio-unit-tests.sh | 50f377f1bffb16dbfe20b0bd1cba7794a2b0da81 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-unit-tests/9588/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Rigdon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 50f377f1bffb16dbfe20b0bd1cba7794a2b0da81 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-unit-tests/9588/) | `/test istio-unit-tests`
prow/istio-presubmit.sh | 50f377f1bffb16dbfe20b0bd1cba7794a2b0da81 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6366/istio-presubmit/13815/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

gyliu513 commented:
/retest

sdake commented:
/lgtm

Nice work Ryan!

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6366#pullrequestreview-131847348" title="LGTM">gyliu513</a>*, *<a href="https://github.com/istio/istio/pull/6366#" title="Author self-approved">Rigdon</a>*, *<a href="https://github.com/istio/istio/pull/6366#issuecomment-401617613" title="LGTM">sdake</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[install/OWNERS](https://github.com/istio/istio/blob/master/install/OWNERS)~~ [gyliu513,sdake]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
